index/miner/onchain.go

```diff
@@ -0,0 +1,144 @@
+package miner
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/textileio/filecoin/lotus/types"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+const (
+	fullRefreshThreshold = 100
+	heightOffset         = uint64(20)
+)
+
+// updateOnChainIndex updates on-chain index information when a new tipset
+func (mi *MinerIndex) updateOnChainIndex(new *types.TipSet) error {
+	log.Info("updating on-chain index...")
+	var chainIndex ChainIndex
+	newtsk := types.NewTipSetKey(new.Cids...)
+	ts, err := mi.store.LoadAndPrune(mi.ctx, newtsk, &chainIndex)
+	if err != nil {
+		return err
+	}
+	if chainIndex.Power == nil {
+		chainIndex.Power = make(map[string]Power)
+	}
+	mctx := context.Background()
+	start := time.Now()
+	log.Infof("current state height %d, new tipset height %d", chainIndex.LastUpdated, new.Height)
+	if new.Height-chainIndex.LastUpdated > fullRefreshThreshold {
+		mctx, _ = tag.New(mctx, tag.Insert(metricRefreshType, "full"))
+		new, err = mi.api.ChainGetTipSetByHeight(mi.ctx, new.Height-heightOffset, new)
+		if err != nil {
+			return err
+		}
+		if err := fullRefresh(mi.ctx, mi.api, &chainIndex); err != nil {
+			return fmt.Errorf("error doing full refresh: %s", err)
+		}
+
+	} else {
+		mctx, _ = tag.New(mctx, tag.Insert(metricRefreshType, "delta"))
+		if err := deltaRefresh(mi.ctx, mi.api, &chainIndex, *ts, new); err != nil {
+			return fmt.Errorf("error doing delta refresh: %s", err)
+		}
+	}
+	chainIndex.LastUpdated = new.Height
+	stats.Record(mctx, mRefreshDuration.M(int64(time.Since(start).Milliseconds())))
+
+	mi.lock.Lock()
+	mi.index.Chain = chainIndex
+	mi.lock.Unlock()
+
+	if err := mi.store.Save(mi.ctx, types.NewTipSetKey(new.Cids...), chainIndex); err != nil {
```

Comment:
Now that we have the new `chainIndex` which got built from the last saved state, up to the new Tipset. we call `Save` which basically saves this new state that corresponds to the `new` Tipset mark.

On future updates that doesn't fork from `new`, the `LoadAndPrune` call will return this state if can be used to move forward. And the cycle continues.